File: src/main/java/ch/itds/taglib/asset/pipeline/thymeleaf/AssetExpressionFactory.java
package ch.itds.taglib.asset.pipeline.thymeleaf;

import org.thymeleaf.context.IExpressionContext;
import org.thymeleaf.expression.IExpressionObjectFactory;

import java.util.Collections;
import java.util.Set;

public class AssetExpressionFactory implements IExpressionObjectFactory {

    private static final String EVALUATION_VARIABLE_NAME = "asset";

    private static final Set<String> ALL_EXPRESSION_OBJECT_NAMES = Collections.unmodifiableSet(
            Collections.singleton(EVALUATION_VARIABLE_NAME)
    );

    @Override
    public Set<String> getAllExpressionObjectNames() {
        return ALL_EXPRESSION_OBJECT_NAMES;
    }

    @Override
    public Object buildObject(IExpressionContext context, String expressionObjectName) {
        if (EVALUATION_VARIABLE_NAME.equals(expressionObjectName)) {
            return new AssetExpressionObject();
        }
        return null;
    }

    @Override
    public boolean isCacheable(String expressionObjectName) {
        return EVALUATION_VARIABLE_NAME.equals(expressionObjectName);
    }
}

File: src/main/java/com/dra11y/assetThymeleafTaglib/expression/AssetExpressionFactory.kt
package com.dra11y.assetThymeleafTaglib.expression

import org.springframework.context.annotation.Import
import org.springframework.stereotype.Component
import org.thymeleaf.context.IExpressionContext
import org.thymeleaf.expression.IExpressionObjectFactory

@Component
@Import(AssetExpression::class)
class AssetExpressionFactory(private val assetExpression: AssetExpression) : IExpressionObjectFactory {
    override fun getAllExpressionObjectNames(): Set<String> = ALL_EXPRESSION_OBJECT_NAMES

    override fun buildObject(context: IExpressionContext, expressionObjectName: String): Any? =
        if (EVALUATION_VARIABLE_NAME == expressionObjectName) {
            assetExpression
        } else null

    override fun isCacheable(expressionObjectName: String): Boolean =
        EVALUATION_VARIABLE_NAME == expressionObjectName

    companion object {
        private const val EVALUATION_VARIABLE_NAME = "asset"
        private val ALL_EXPRESSION_OBJECT_NAMES = setOf(EVALUATION_VARIABLE_NAME)
    }
}
